feature(engine): add activityIdIn filter to HistoricProcessInstanceQuery

venetrius opened the pull request:
related to: #4618

tasso94 reviewed (changes requested):
* You are only testing the use case with a single activityId but not with multiple ids.
* Can we test multiple activities in different scopes with and without incidents?
* ❓ Does it make sense to copy some of these tests over to `HistoricProcessInstanceTest` (please scroll in the code excerpt to see all the tests)
https://github.com/camunda/camunda-bpm-platform/blob/1c5a1265080faaecd6d542d889cac9230033ba84/engine/src/test/java/org/camunda/bpm/engine/test/api/runtime/ProcessInstanceQueryTest.java#L1997-L2192
  * One candidate would be passing `null` but also some others might make sense.

yanavasileva commented on engine-rest/engine-rest/src/main/java/org/camunda/bpm/engine/rest/dto/history/HistoricProcessInstanceQueryDto.java:
```diff
@@ -336,6 +337,11 @@ public void setActiveActivityIdIn(List<String> activeActivityIdIn) {
   }
 
+  @CamundaQueryParam(value = "activityIdIn", converter = StringListConverter.class)
+  public void setactivityIdIn(List<String> activityIdIn) {
```
```suggestion
  public void setActivityIdIn(List<String> activityIdIn) {
```

yanavasileva commented on engine/src/main/resources/org/camunda/bpm/engine/impl/mapping/entity/HistoricProcessInstance.xml:
```diff
@@ -626,6 +636,21 @@
               )
             </if>
 
+            <if test="query.activityIds != null and query.activityIds.length > 0">
```
❓ Do we need to consider adding some indexes here? To ensure performance for the modification. Or the existing indexes will be good enough?

tasso94 replied:
Let's be careful with indices on history tables. I would only add them when we get feedback. Adding a new index for existing databases with tables that have huge amounts of data is a costly operation, and it slows down process execution since history insertion is slower.

yanavasileva reviewed (commented):
❓ Do we need to test call activities? I mean an incident in a call activity.

yanavasileva reviewed (changes requested):
✍️ To add to Tassilo's suggestions further.
We need to have tests for filtering historic instances where some of them have been already completed/deleted. 
Similar to what was done in https://github.com/camunda/camunda-bpm-platform/pull/4624 test coverage. Since that will be the usual case to do modification.

venetrius commented:
> * You are only testing the use case with a single activityId but not with multiple ids.
> * Can we test multiple activities in different scopes with and without incidents?
> * ❓ Does it make sense to copy some of these tests over to `HistoricProcessInstanceTest` (please scroll in the code excerpt to see all the tests)
>   https://github.com/camunda/camunda-bpm-platform/blob/1c5a1265080faaecd6d542d889cac9230033ba84/engine/src/test/java/org/camunda/bpm/engine/test/api/runtime/ProcessInstanceQueryTest.java#L1997-L2192
>   
>   
>   * One candidate would be passing `null` but also some others might make sense.

- Added test when ActivityIdIn has multiple entries
- Added tests with  multiple activities in different scopes with and without incidents
- Copy pasted the tests that are testing activityIdIn filter, it makes sense as we expect the two filter work in the same way in the 2 query types.  **Three of the copied test fails** as queries do not yield the expected results, I added `@Ignore` annotation to them, but will come back later to see. If you have any insight in the mean time that  would be most welcomed.

venetrius commented:
> ✍️ To add to Tassilo's suggestions further. We need to have tests for filtering historic instances where some of them have been already completed/deleted. Similar to what was done in #4624 test coverage. Since that will be the usual case to do modification.

Added test when some instances are deleted and completed

venetrius commented:
ActivityIdIn filter returns different processInstances for runtimeQuery and HistoricQuery  in special cases around  `AsyncBefore`, `AsyncAfter`, and compensation.
Updated the [OpenApi doc](https://github.com/camunda/camunda-bpm-platform/pull/4619/files#diff-7f9daf065d646ff9aa460dc862e7f5bfaa3769056bce9d9f3cf2d37743cd14f1),  [HistoricProcessInstanceQuery.java](https://github.com/camunda/camunda-bpm-platform/pull/4619/files#diff-487ef87af3a819a15c779807ce1ba7507dd36abace50d7abdcd89017017d1ba6) and [relevant unit tests](https://github.com/camunda/camunda-bpm-platform/pull/4619/files#diff-4ba32283c1607b311dea3927d651658ff7a751f627cd9f90f62d05300002de4c) (`testQueryByAsyncBeforeActivityId`, `testQueryByAsyncAfterActivityId`, `testQueryByActivityIdDuringCompensation`) to document it.
